File: secator/tasks/testssl.py
import json
import os
from datetime import datetime

from secator.config import CONFIG
from secator.decorators import task
from secator.output_types import Vulnerability, Certificate, Error, Info, Ip, Tag
from secator.definitions import (PROXY, HOST, USER_AGENT, HEADER, OUTPUT_PATH,
                                CERTIFICATE_STATUS_UNKNOWN, CERTIFICATE_STATUS_TRUSTED, CERTIFICATE_STATUS_REVOKED,
                                TIMEOUT)
from secator.tasks._categories import Command, OPTS


@task()
class testssl(Command):
    """SSL/TLS security scanner, including ciphers, protocols and cryptographic flaws."""
    cmd = 'testssl.sh'
    tags = ['dns', 'recon', 'tls']
    input_types = [HOST]
    input_flag = None
    file_flag = '-iL'
    file_eof_newline = True
    version_flag = ''
    opt_prefix = '--'
    opts = {
        'verbose': {'is_flag': True, 'default': False, 'internal': True, 'display': True, 'help': 'Record all SSL/TLS info, not only critical info'},  # noqa: E501
        'parallel': {'is_flag': True, 'default': False, 'help': 'Test multiple hosts in parallel'},
        'warnings': {'type': str, 'default': None, 'help': 'Set to "batch" to stop on errors, and "off" to skip errors and continue'},  # noqa: E501
        'ids_friendly': {'is_flag': True, 'default': False, 'help': 'Avoid IDS blocking by skipping a few vulnerability checks'},  # noqa: E501
        'hints': {'is_flag': True, 'default': False, 'help': 'Additional hints to findings'},
        'server_defaults': {'is_flag': True, 'default': False, 'help': 'Displays the server default picks and certificate info'},  # noqa: E501
    }
    meta_opts = {
        PROXY: OPTS[PROXY],
        USER_AGENT: OPTS[USER_AGENT],
        HEADER: OPTS[HEADER],
        TIMEOUT: OPTS[TIMEOUT],
    }
    opt_key_map = {
        PROXY: 'proxy',
        USER_AGENT: 'user-agent',
        HEADER: 'reqheader',
        TIMEOUT: 'connect-timeout',
        'ipv6': '-6',
    }
    output_types = [Certificate, Vulnerability, Ip, Tag]
    proxy_http = True
    proxychains = False
    proxy_socks5 = False
    profile = 'io'
    install_pre = {
        'apk': ['hexdump', 'coreutils', 'procps'],
        'pacman': ['util-linux'],
        '*': ['bsdmainutils']
    }
    install_version = 'v3.2.0'
    install_cmd = (
        f'git clone --depth 1 --single-branch -b [install_version] https://github.com/drwetter/testssl.sh.git {CONFIG.dirs.share}/testssl.sh_[install_version] || true && '  # noqa: E501
        f'ln -sf {CONFIG.dirs.share}/testssl.sh_[install_version]/testssl.sh {CONFIG.dirs.bin}'
    )

    @staticmethod
    def on_cmd(self):
        output_path = self.get_opt_value(OUTPUT_PATH)
        if not output_path:
            output_path = f'{self.reports_folder}/.outputs/{self.unique_name}.json'
        self.output_path = output_path
        self.cmd += f' --jsonfile {self.output_path}'

        # Hack because target needs to be the last argument in testssl.sh
        if len(self.inputs) == 1:
            target = self.inputs[0]
            self.cmd = self.cmd.replace(f' {target}', '')
            self.cmd += f' {target}'

    @staticmethod
    def on_cmd_done(self):
        if not os.path.exists(self.output_path):
            yield Error(message=f'Could not find JSON results in {self.output_path}')
            return
        yield Info(message=f'JSON results saved to {self.output_path}')

        verbose = self.get_opt_value('verbose')
        with open(self.output_path, 'r') as f:
            data = json.load(f)
            bad_cyphers = {}
            retrieved_certificates = {}
            ignored_item_ids = ["scanTime", "overall_grade", "DNS_CAArecord"]
            ip_addresses = []
            host_to_ips = {}

            for item in data:
                host, ip = tuple(item['ip'].split('/'))
                id = item['id']
                # port = item['port']
                finding = item['finding']
                severity = item['severity'].lower()
                cwe = item.get('cwe')
                vuln_tags = ['ssl', 'tls']
                if cwe:
                    vuln_tags.append(cwe)

                # Skip ignored items
                if id.startswith(tuple(ignored_item_ids)):
                    continue

                # Add IP to address pool
                host_to_ips.setdefault(host, []).append(ip)
                if ip not in ip_addresses:
                    ip_addresses.append(ip)
                    yield Ip(
                        host=host,
                        ip=ip,
                        alive=True
                    )

                # Process errors
                if id.startswith("scanProblem"):
                    yield Error(message=finding)

                # Process bad ciphers
                elif id.startswith('cipher-'):
                    splited_item = item["finding"].split(" ")
                    concerned_protocol = splited_item[0]
                    bad_cypher = splited_item[-1]
                    bad_cyphers.setdefault(ip, {}).setdefault(concerned_protocol, []).append(bad_cypher)  # noqa: E501

                # Process certificates
                elif id.startswith('cert_') or id.startswith('cert '):
                    retrieved_certificates.setdefault(ip, []).append(item)

                # Process intermediate certificates
                elif id.startswith('intermediate_cert_'):
                    # TODO: implement this
                    pass

                # If info or ok, create a tag only if 'verbose' option is set
                elif severity in ['info', 'ok']:
                    if not verbose:
                        continue
                    yield Tag(
                        name=f'SSL/TLS [{id}]',
                        match=host,
                        extra_data={
                            'type': id,
                            'finding': finding,
                        }
                    )

                # Create vulnerability
                else:
                    if id in ['TLS1', 'TLS1_1']:
                        human_name = f'SSL/TLS deprecated protocol offered: {id}'
                    else:
                        human_name = f'SSL/TLS {id}: {finding}'
                    yield Vulnerability(
                        name=human_name,
                        matched_at=host,
                        ip=ip,
                        tags=vuln_tags,
                        severity=severity,
                        confidence='high',
                        extra_data={
                            'id': id,
                            'finding': finding
                        }
                    )

            # Creating vulnerability for the deprecated ciphers
            for ip, protocols in bad_cyphers.items():
                for protocol, cyphers in protocols.items():
                    yield Vulnerability(
                        name=f'SSL/TLS vulnerability ciphers for {protocol} deprecated',
                        matched_at=ip,
                        ip=ip,
                        confidence='high',
                        severity='low',
                        extra_data={
                            'cyphers': cyphers
                        }
                    )

            # Creating certificates for each founded target
            host_to_ips = {k: set(v) for k, v in host_to_ips.items()}
            for ip, certs in retrieved_certificates.items():
                host = [k for k, v in host_to_ips.items() if ip in v][0]
                cert_data = {
                    'host': host,
                    'ip': ip,
                    'fingerprint_sha256': None,
                    'subject_cn': None,
                    'subject_an': None,
                    'not_before': None,
                    'not_after': None,
                    'issuer_cn': None,
                    'self_signed': None,
                    'trusted': None,
                    'status': None,
                    'keysize': None,
                    'serial_number': None,
                }
                for cert in certs:
                    host = [k for k, v in host_to_ips.items() if ip in v][0]
                    id = cert['id']
                    finding = cert['finding']

                    if id.startswith('cert_crlDistributionPoints') and finding != '--':
                        # TODO not implemented, need to find a certificate that is revoked by CRL
                        cert_data['status'] = CERTIFICATE_STATUS_UNKNOWN

                    if id.startswith('cert_ocspRevoked'):
                        if finding.startswith('not revoked'):
                            cert_data['status'] = CERTIFICATE_STATUS_TRUSTED
                        else:
                            cert_data['status'] = CERTIFICATE_STATUS_REVOKED

                    if id.startswith('cert_fingerprintSHA256'):
                        cert_data['fingerprint_sha256'] = finding

                    if id.startswith('cert_commonName'):
                        cert_data['subject_cn'] = finding

                    if id.startswith('cert_subjectAltName'):
                        cert_data['subject_an'] = finding.split(" ")

                    if id.startswith('cert_notBefore'):
                        cert_data['not_before'] = datetime.strptime(finding, "%Y-%m-%d %H:%M")

                    if id.startswith('cert_notAfter'):
                        cert_data['not_after'] = datetime.strptime(finding, "%Y-%m-%d %H:%M")

                    if id.startswith('cert_caIssuers'):
                        cert_data['issuer_cn'] = finding

                    if id.startswith('cert_chain_of_trust'):
                        cert_data['self_signed'] = 'self signed' in finding

                    if id.startswith('cert_chain_of_trust'):
                        cert_data['trusted'] = finding.startswith('passed')

                    if id.startswith('cert_keySize'):
                        cert_data['keysize'] = int(finding.split(" ")[1])

                    if id.startswith('cert_serialNumber'):
                        cert_data['serial_number'] = finding

                    if id.startswith('cert ') and finding.startswith('-----BEGIN CERTIFICATE-----'):
                        cert_data['raw_value'] = finding

                # For the following attributes commented, it's because at the time of writting it
                # I did not found the value inside the result of testssl
                cert = Certificate(
                    **cert_data
                    # issuer_dn='',
                    # issuer='',
                    # TODO: delete the ciphers attribute from certificate outputType
                    # ciphers=None,
                    # TODO: need to find a way to retrieve the parent certificate,
                    # parent_certificate=None,
                )
                yield cert
                if cert.is_expired():
                    yield Vulnerability(
                        name='SSL certificate expired',
                        provider='testssl',
                        description='The SSL certificate is expired. This can easily lead to domain takeovers',
                        matched_at=host,
                        ip=ip,
                        tags=['ssl', 'tls'],
                        severity='medium',
                        confidence='high',
                        extra_data={
                            'id': id,
                            'expiration_date': Certificate.format_date(cert.not_after)
                        }
                    )

File: rust/crates/secator-tasks/src/testssl.rs
//! testssl.sh — TLS/SSL scanner (Python `secator/tasks/testssl.py`).
//!
//! testssl emits a JSON file via `--jsonfile <path>`. We inject the path inside
//! `on_cmd` (under the run's `.outputs/`), then parse it in `on_cmd_done`:
//!   * `cipher-*`            → grouped per `(ip, protocol)` → Vulnerability(low)
//!   * `cert_*`              → grouped per ip → Certificate
//!   * `id == scanProblem*`  → Warning
//!   * `id == engine_problem*` → Warning
//!   * severity ∈ {info, ok} → Tag(category=info, name=ssl_tls) when verbose
//!   * everything else with non-info/ok severity → Vulnerability
//!
//! Mirrors `Python ::on_cmd_done` field-by-field. Skip-id list (`scanTime`,
//! `overall_grade`, `DNS_CAArecord`) is preserved.

use std::fs;

use secator_model::{Certificate, Info, Ip, Map, OutputItem, Tag, Vulnerability, Warning};
use secator_options::{FileMode, InputWiring, KeyMap, OptSchema, OptSpec, OptType, SingleMode};
use secator_runner::{empty_output_maps, HookCtx, HookRegistry, ValidatorRegistry};
use serde_json::Value;

use crate::TaskSpec;

pub static SPEC: TaskSpec = TaskSpec {
    name: "testssl",
    description: "SSL/TLS security scanner — ciphers, protocols, crypto flaws.",
    cmd: "testssl.sh",
    input_types: &["host", "host_port", "url", "ip"],
    output_types: &["certificate", "vulnerability", "ip", "tag"],
    tags: &["dns", "recon", "tls"],
    // No `--json` on stdout — testssl writes a JSON file at `--jsonfile <path>`.
    // We don't stream stdout JSON; everything happens in `on_cmd_done`.
    json_flag: None,
    input_wiring: InputWiring { single: SingleMode::Arg, file: FileMode::Flag("-iL") },
    item_loaders: &[],
    input_chunk_size: 1,
    on_json_loaded: None,
    on_regex_loaded: None,
    output_maps: empty_output_maps,
    discriminator: None,
    hooks: HOOKS,
    validators: ValidatorRegistry::EMPTY,
    schema: build_schema,
    install: secator_runner::InstallSpec {
        version: Some("v3.2.0"),
        cmd: Some("git clone --depth 1 --single-branch -b [install_version] https://github.com/drwetter/testssl.sh.git $HOME/.local/share/testssl.sh_[install_version] || true && ln -sf $HOME/.local/share/testssl.sh_[install_version]/testssl.sh $HOME/.local/bin"),
        github_handle: Some("testssl/testssl.sh"),
        github_bin: false,
        cmd_pre: &[
            ("apk", &["hexdump", "coreutils", "procps", "bash"]),
            ("pacman", &["util-linux", "bash"]),
            ("*", &["bsdmainutils", "bash"]),
        ],
        ..secator_runner::InstallSpec::EMPTY
    },
    proxy_caps: secator_runner::ProxyCaps::NONE,
    encoding: "utf-8",
    ignore_return_code: false,
    requires_sudo: false,
    default_inputs: None,
};

static HOOKS: HookRegistry = HookRegistry {
    before_init: &[before_init_record_state],
    on_cmd: &[on_cmd_inject_jsonfile],
    on_cmd_done: &[on_cmd_done_parse],
    ..HookRegistry::EMPTY
};

fn before_init_record_state(
    ctx: &mut HookCtx,
    runner: &mut secator_runner::CommandRunner,
) {
    if let Some(rf) = &runner.reports_folder {
        ctx.state.insert("testssl:reports".into(), rf.to_string_lossy().into_owned());
    }
    if runner.opts.get("verbose").map(|v| v.as_str()) == Some("true") {
        ctx.state.insert("testssl:verbose".into(), "1".into());
    }
}

/// Python `on_cmd`: append `--jsonfile <reports>/.outputs/testssl.json` to the
/// cmd. The target also needs to be the LAST positional in testssl.sh —
/// we don't reorder here because `CommandRunner::build_cmd` already places
/// inputs at the end.
fn on_cmd_inject_jsonfile(ctx: &mut HookCtx, cmd: &mut String) {
    let reports = ctx.state.get("testssl:reports").cloned().unwrap_or_default();
    let outputs_dir = if reports.is_empty() {
        String::from("/tmp")
    } else {
        format!("{reports}/.outputs")
    };
    let _ = fs::create_dir_all(&outputs_dir);
    let path = format!("{outputs_dir}/testssl.json");
    let quoted = shell_words::quote(&path).into_owned();
    cmd.push_str(&format!(" --jsonfile {quoted}"));
    ctx.state.insert("testssl:output_path".into(), path);
}

fn on_cmd_done_parse(ctx: &mut HookCtx) -> Vec<OutputItem> {
    let path = match ctx.state.get("testssl:output_path") {
        Some(p) => p.clone(),
        None => return Vec::new(),
    };
    let body = match fs::read_to_string(&path) {
        Ok(s) => s,
        Err(_) => {
            // testssl exits cleanly against hosts without TLS / with hardened
            // configs — the JSON file just never gets written. Treat as a
            // Warning (not an Error) so the run status stays SUCCESS.
            return vec![OutputItem::Warning(Warning {
                message: format!("Could not find JSON results in {path}"),
                ..Default::default()
            })];
        }
    };
    let arr: Vec<Value> = match serde_json::from_str(&body) {
        Ok(Value::Array(a)) => a,
        Ok(_) | Err(_) => return Vec::new(),
    };
    let verbose = ctx.state.contains_key("testssl:verbose");
    let ignored_prefixes = ["scanTime", "overall_grade", "DNS_CAArecord"];

    let mut out: Vec<OutputItem> = vec![OutputItem::Info(Info {
        message: format!("JSON results saved to {path}"),
        ..Default::default()
    })];

    // Two accumulators: per-ip bad ciphers (id starts with `cipher-`) and per-ip
    // certificate fields (id starts with `cert_` / `cert ` / `intermediate_cert_`).
    use std::collections::BTreeMap;
    let mut bad_cyphers: BTreeMap<String, BTreeMap<String, Vec<String>>> = BTreeMap::new();
    let mut cert_items: BTreeMap<String, Vec<Value>> = BTreeMap::new();
    let mut host_to_ips: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let mut seen_ips: Vec<String> = Vec::new();

    for item in &arr {
        let ip_field = item.get("ip").and_then(|v| v.as_str()).unwrap_or("");
        let (host, ip) = ip_field
            .split_once('/')
            .map(|(a, b)| (a.to_string(), b.to_string()))
            .unwrap_or((ip_field.to_string(), String::new()));
        let id = item.get("id").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let finding = item.get("finding").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let severity = item.get("severity").and_then(|v| v.as_str()).unwrap_or("").to_lowercase();
        let cwe = item.get("cwe").and_then(|v| v.as_str()).unwrap_or("").to_string();

        if ignored_prefixes.iter().any(|p| id.starts_with(p)) {
            continue;
        }

        host_to_ips.entry(host.clone()).or_default().push(ip.clone());
        if !ip.is_empty() && !seen_ips.contains(&ip) {
            seen_ips.push(ip.clone());
            out.push(OutputItem::Ip(Ip {
                host: host.clone(),
                ip: ip.clone(),
                alive: true,
                ..Default::default()
            }));
        }

        if id.starts_with("scanProblem") || id.starts_with("engine_problem") {
            out.push(OutputItem::Warning(Warning { message: finding, ..Default::default() }));
            continue;
        }
        if id.starts_with("cipher-") {
            let parts: Vec<&str> = finding.split_whitespace().collect();
            if parts.len() >= 2 {
                let protocol = parts[0].to_string();
                let bad = parts.last().unwrap().to_string();
                bad_cyphers
                    .entry(ip.clone())
                    .or_default()
                    .entry(protocol)
                    .or_default()
                    .push(bad);
            }
            continue;
        }
        if id.starts_with("cert_") || id.starts_with("cert ") {
            cert_items.entry(ip.clone()).or_default().push(item.clone());
            continue;
        }
        if id.starts_with("intermediate_cert_") {
            continue;
        }
        if severity == "info" || severity == "ok" {
            if !verbose {
                continue;
            }
            let mut extra: Map = Map::new();
            extra.insert("subtype".into(), Value::String(id.clone()));
            out.push(OutputItem::Tag(Tag {
                category: "info".into(),
                name: "ssl_tls".into(),
                match_: host.clone(),
                value: finding,
                extra_data: extra,
                ..Default::default()
            }));
            continue;
        }
        // Anything else with a real severity ⇒ Vulnerability.
        let human_name = if id == "TLS1" || id == "TLS1_1" {
            format!("SSL/TLS deprecated protocol offered: {id}")
        } else {
            format!("SSL/TLS {id}")
        };
        let mut extra: Map = Map::new();
        extra.insert("id".into(), Value::String(id.clone()));
        extra.insert("finding".into(), Value::String(finding));
        let mut tags = vec!["ssl".to_string(), "tls".to_string()];
        if !cwe.is_empty() {
            tags.push(cwe);
        }
        out.push(OutputItem::Vulnerability(Vulnerability {
            name: human_name,
            matched_at: host.clone(),
            ip: ip.clone(),
            tags,
            severity,
            confidence: "high".into(),
            extra_data: extra,
            ..Default::default()
        }));
    }

    // Emit a low-severity Vulnerability per (ip, protocol) for deprecated ciphers.
    for (ip, by_proto) in &bad_cyphers {
        for (protocol, cyphers) in by_proto {
            let mut extra: Map = Map::new();
            extra.insert(
                "cyphers".into(),
                Value::Array(cyphers.iter().map(|s| Value::String(s.clone())).collect()),
            );
            out.push(OutputItem::Vulnerability(Vulnerability {
                name: format!("SSL/TLS vulnerability ciphers for {protocol} deprecated"),
                matched_at: ip.clone(),
                ip: ip.clone(),
                severity: "low".into(),
                confidence: "high".into(),
                extra_data: extra,
                ..Default::default()
            }));
        }
    }

    // Reverse-lookup: ip → host (use the first host that mentions this ip).
    let ip_to_host = |needle: &str| -> String {
        for (h, ips) in &host_to_ips {
            if ips.iter().any(|i| i == needle) {
                return h.clone();
            }
        }
        String::new()
    };

    for (ip, certs) in &cert_items {
        let mut cert = Certificate {
            host: ip_to_host(ip),
            ip: ip.clone(),
            ..Default::default()
        };
        for c in certs {
            let cid = c.get("id").and_then(|v| v.as_str()).unwrap_or("");
            let fnd = c.get("finding").and_then(|v| v.as_str()).unwrap_or("").to_string();
            if cid.starts_with("cert_crlDistributionPoints") && fnd != "--" {
                cert.status = "Unknown".into();
            }
            if cid.starts_with("cert_ocspRevoked") {
                cert.status = if fnd.starts_with("not revoked") {
                    "Trusted".into()
                } else {
                    "Revoked".into()
                };
            }
            if cid.starts_with("cert_fingerprintSHA256") {
                cert.fingerprint_sha256 = fnd.clone();
            }
            if cid.starts_with("cert_commonName") {
                cert.subject_cn = fnd.clone();
            }
            if cid.starts_with("cert_subjectAltName") {
                cert.subject_an = fnd.split_whitespace().map(str::to_string).collect();
            }
            if cid.starts_with("cert_notBefore") {
                cert.not_before = Some(fnd.clone());
            }
            if cid.starts_with("cert_notAfter") {
                cert.not_after = Some(fnd.clone());
            }
            if cid.starts_with("cert_caIssuers") {
                cert.issuer_cn = fnd.clone();
            }
            if cid.starts_with("cert_chain_of_trust") {
                cert.self_signed = fnd.contains("self signed");
                cert.trusted = fnd.starts_with("passed");
            }
            if cid.starts_with("cert_keySize") {
                if let Some(num) = fnd.split_whitespace().nth(1) {
                    cert.keysize = num.parse().ok();
                }
            }
            if cid.starts_with("cert_serialNumber") {
                cert.serial_number = fnd.clone();
            }
            if cid.starts_with("cert ") && fnd.starts_with("-----BEGIN CERTIFICATE-----") {
                cert.raw_value = fnd.clone();
            }
        }
        out.push(OutputItem::Certificate(cert));
    }

    out
}

fn build_schema() -> OptSchema {
    let mut s = OptSchema { opt_prefix: "--", ..OptSchema::default() };
    s.meta_opts = crate::meta_opts::opts_recon();
    crate::meta_opts::apply_config_defaults(&mut s.meta_opts);
    // Python `opt_key_map`: rename a couple of common opts to testssl's flag names.
    s.key_map.insert("proxy".into(), KeyMap::Flag("proxy".into()));
    s.key_map.insert("user_agent".into(), KeyMap::Flag("user-agent".into()));
    s.key_map.insert("header".into(), KeyMap::Flag("reqheader".into()));
    s.key_map.insert("timeout".into(), KeyMap::Flag("connect-timeout".into()));
    for k in ["delay", "rate_limit", "retries", "threads"] {
        s.key_map.insert(k.into(), KeyMap::NotSupported);
    }
    s.opts = vec![
        OptSpec {
            name: "verbose",
            ty: OptType::Bool,
            short: None,
            is_flag: true,
            default: None,
            help: "Record all SSL/TLS info, not only critical issues",
            internal: true,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "parallel",
            ty: OptType::Bool,
            short: None,
            is_flag: true,
            default: None,
            help: "Test multiple hosts in parallel",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "warnings",
            ty: OptType::Str,
            short: None,
            is_flag: false,
            default: None,
            help: "Set to 'batch' to stop on errors, 'off' to skip them",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "ids_friendly",
            ty: OptType::Bool,
            short: None,
            is_flag: true,
            default: None,
            help: "Avoid IDS blocking by skipping some vulnerability checks",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "hints",
            ty: OptType::Bool,
            short: None,
            is_flag: true,
            default: None,
            help: "Additional hints to findings",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "server_defaults",
            ty: OptType::Bool,
            short: None,
            is_flag: true,
            default: None,
            help: "Display the server default picks and certificate info",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
    ];
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_minimal_testssl_json() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("t.json");
        let body = r#"[
            {"id": "TLS1", "ip": "example.com/93.184.216.34", "port": "443",
             "severity": "HIGH", "finding": "offered", "cwe": "CWE-310"},
            {"id": "cipher-tls1_3-1", "ip": "example.com/93.184.216.34", "port": "443",
             "severity": "MEDIUM", "finding": "TLSv1.3 weak NULL"},
            {"id": "cert_commonName", "ip": "example.com/93.184.216.34", "port": "443",
             "severity": "ok", "finding": "www.example.com"},
            {"id": "scanProblem-1", "ip": "example.com/93.184.216.34", "port": "443",
             "severity": "WARN", "finding": "host is down"}
        ]"#;
        std::fs::write(&path, body).unwrap();
        let mut ctx = HookCtx::default();
        ctx.state.insert("testssl:output_path".into(), path.to_string_lossy().into_owned());
        let out = on_cmd_done_parse(&mut ctx);

        // Expect: Info (saved), Ip, Vulnerability (TLS1), Warning (scanProblem),
        // Certificate, Vulnerability (deprecated ciphers).
        let has_vuln_tls = out.iter().any(|i| matches!(i, OutputItem::Vulnerability(v) if v.name.contains("TLS1")));
        let has_cert = out.iter().any(|i| matches!(i, OutputItem::Certificate(_)));
        let has_warning = out.iter().any(|i| matches!(i, OutputItem::Warning(_)));
        let has_ip = out.iter().any(|i| matches!(i, OutputItem::Ip(_)));
        let has_bad_cypher_vuln = out
            .iter()
            .any(|i| matches!(i, OutputItem::Vulnerability(v) if v.name.contains("deprecated")));
        assert!(has_vuln_tls && has_cert && has_warning && has_ip && has_bad_cypher_vuln);
    }

    /// When testssl exits without writing its JSON file (target has no TLS,
    /// hardened target, etc.) we must emit a Warning, not an Error — the run
    /// status should stay SUCCESS.
    #[test]
    fn missing_json_file_emits_warning_not_error() {
        let mut ctx = HookCtx::default();
        ctx.state.insert(
            "testssl:output_path".into(),
            "/nonexistent/path/testssl.json".into(),
        );
        let out = on_cmd_done_parse(&mut ctx);
        assert_eq!(out.len(), 1);
        assert!(
            matches!(&out[0], OutputItem::Warning(w) if w.message.contains("Could not find JSON results")),
            "expected a Warning for missing JSON, got {:?}",
            out[0]
        );
        // Explicitly assert *no* Error variant.
        assert!(!out.iter().any(|i| matches!(i, OutputItem::Error(_))));
    }
}
